### ISSUE TITLE
Debug.Assert failure in NullableWalker.TakeIncrementalSnapshot

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#47102]
Discovered while triaging #46536

**Version Used**: f6edf2868c49ec9d3d494c26cd6ad9f710355d2a

**Steps to Reproduce**:

```cs
        [Fact, WorkItem(46536, "https://github.com/dotnet/roslyn/issues/46536")]
        public void MultiplePathsToNode_SwitchDispatch_02()
        {
            var source = @"
class Program
{
    static int Main()
    {
        return (false, 0) switch
        {
            (false, 0) => 0,
            (false, 1) => 0, 
            (false or true, 2) => 0
        };
    }
}
";
            var verifier = CompileAndVerify(source, expectedReturnCode: 0, options: TestOptions.DebugExe.WithNullableContextOptions(NullableContextOptions.Disable));
            verifier.VerifyDiagnostics();
        }
```

**Expected Behavior**: We get a warning in VerifyDiagnostics about non-exhaustiveness of the switch expression.

**Actual Behavior**:

We fail the following assertion:

https://github.com/dotnet/roslyn/blob/1ffea3e34ad7da15bcc93704e614c16096fe017a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L1284

With the following stack trace:
```
The active test run was aborted. Reason: Test host process crashed : Process terminated. Assertion failed.
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.TakeIncrementalSnapshot(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 1284
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node, Boolean expressionIsRead) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2773
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2760
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitBinaryPattern(BoundBinaryPattern node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker_Patterns.cs:line 95
   at Microsoft.CodeAnalysis.CSharp.BoundBinaryPattern.Accept(BoundTreeVisitor visitor) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 7577
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 149
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitWithStackGuard(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 359
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitAlways(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 343
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 318
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node, Boolean expressionIsRead) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2774
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2760
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitSubpattern(BoundSubpattern node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker_Patterns.cs:line 43
   at Microsoft.CodeAnalysis.CSharp.BoundSubpattern.Accept(BoundTreeVisitor visitor) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 7507
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 149
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitWithStackGuard(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 359
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitAlways(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 343
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 318
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node, Boolean expressionIsRead) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2774
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2760
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitAndUnsplitAll[T](ImmutableArray`1 nodes) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 1579
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitRecursivePattern(BoundRecursivePattern node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker_Patterns.cs:line 50
   at Microsoft.CodeAnalysis.CSharp.BoundRecursivePattern.Accept(BoundTreeVisitor visitor) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 7438
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 149
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitWithStackGuard(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 359
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitAlways(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 343
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 318
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node, Boolean expressionIsRead) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2774
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2760
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitWithoutDiagnostics(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 1588
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitPatternForRewriting(BoundPattern pattern) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker_Patterns.cs:line 37
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitSwitchExpressionCore(BoundSwitchExpression node, Boolean inferType) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker_Patterns.cs:line 617
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitConvertedSwitchExpression(BoundConvertedSwitchExpression node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker_Patterns.cs:line 563
   at Microsoft.CodeAnalysis.CSharp.BoundConvertedSwitchExpression.Accept(BoundTreeVisitor visitor) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 4572
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 149
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitExpressionWithoutStackGuard(BoundExpression node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 365
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitExpressionWithoutStackGuard(BoundExpression node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2683
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 230
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 209
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitWithStackGuard(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 356
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitAlways(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 343
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 318
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node, Boolean expressionIsRead) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2774
```

FYI @333fred

**Comment by RikkiGibson** [comment]
Actually, have a simpler repro:

```cs
        [Fact, WorkItem(46536, "https://github.com/dotnet/roslyn/issues/46536")]
        public void MultiplePathsToNode_SwitchDispatch_02()
        {
            var source = @"
class Program
{
    static int Main()
    {
        return false switch
        {
            false or true => 0
        };
    }
}
";
            var verifier = CompileAndVerify(source, expectedReturnCode: 0, options: TestOptions.DebugExe.WithNullableContextOptions(NullableContextOptions.Disable));
            verifier.VerifyDiagnostics();
        }
```
